### PR TITLE
Check GRACC for accounting record uploads ( SOFTWARE-2722 )

### PR DESCRIPTION
**Pull request by sthapa**

### CODE REVIEW

**Comment by edquist** [comment]
@sthapa, I get the following error (returned from the `curl` command) when I run the new probe code:
```
curl -q --retry 1 --connect-timeout 10 -XGET 'https://gracc.opensciencegrid.org/q/gracc.osg.raw3-2018.1/_search' -d '{ "query": { "query_string": { "query": "ProbeName:\"condor:fermicloud008.fnal.gov\" AND EndTime:[now-2d TO now]" } }, "sort": [ { "@received": { "order": "desc" } } ],"size": 1 }'
```

```
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index","resource.type":"index_or_alias","resource.id":"gracc.osg.raw3-2018.1","index_uuid":"_na_","index":"gracc.osg.raw3-2018.1"}],"type":"index_not_found_exception","reason":
```

Could it be because it's using an invalid raw index name instead of a summary index?

**Comment by bbockelm** [comment]
Poking around the web interface, it appears the index name is `gracc.osg.raw3-2018.01` while you are testing `gracc.osg.raw3-2018.1`.

Double check with @djw8605 that this isn't a regression on the GRACC side, but perhaps just an oversight in the original code?

**Comment by retzkek** [comment]
Neither of those are quite correct, you should be using `gracc.osg.raw-2018.01`, which is internally an alias for `gracc.osg.raw3-2018.01`, the idea being that if the internal version number (3) needs to change it can be done transparently.

**Comment by bbockelm** [comment]
@retzkek - thanks, that indeed was my next question.

(Aliases confuse me).

**Comment by djw8605** [comment]
Why even use these date specific aliases, and not something like `gracc.osg.raw*` ?

**Comment by retzkek** [comment]
Targeting the specific index(es) you know your data is in will always be faster, potentially much faster depending on the query. The example query above returned in 33 ms against `gracc.osg.raw-2018.01` versus 13660 ms against `gracc.osg.raw-*`. Could be interesting to play with different query types (e.g. explicit range query vs. query_string) to see if that matters, maybe Elasticsearch can better optimize some queries since it keeps some basic field stats at the index level (min, max, etc). Left as an exercise to the reader.

**Comment by brianhlin** [comment]
@edquist, please do not let this hold up the RSV build. Since you have write permissions for this repo, you should be able to push directly to Suchandra's branch and update this PR with the correct query.

**Comment by edquist** [comment]
@bbockelm, @retzkek, thanks for the clarification.  I'll correct the query to use the unversioned raw index, and `YYYY.MM` instead of `YYYY.M`.

**Comment by sthapa** [comment]
Thanks for the fixes Carl.  I missed the zero padding in the index name since I was testing last year.

File: rsv-metrics/libexec/probes/gratia-config-probe [hunk 1]
```diff
@@ -268,77 +268,92 @@ EOF
...
+  my $year = 1900 + $cur_time[5];
+  # perl uses 0 based months
+  my $month = 1 + $cur_time[4];
+  my $gracchost = "https://gracc.opensciencegrid.org";
+  my $graccindex = sprintf "gracc.osg.raw-%04d.%02d", $year, $month;
+  my $curl_cmdline = shellquote(
+    qw(curl -q --retry 1  --connect-timeout 10 -XGET ),
+    "$gracchost/q/$graccindex/_search",
+    "-d",
+    qq({ "query": { "query_string": { "query": "ProbeName:\\"$probename\\" AND EndTime:[now-2d TO now]" } }, "sort": [ { "\@received": { "order": "desc" } } ],"size": 1 })
```
Comment (edquist) [inline rsv-metrics/libexec/probes/gratia-config-probe]: I would feel so much better about this if we could use some library for creating & parsing json ... `perl-JSON`, for instance, is available from the OS repos and we could add it to the rpm requirements.  (Either now or else in a future release...)

Comment (sthapa) [reply rsv-metrics/libexec/probes/gratia-config-probe]: Yeah, I can make the changes to use perl-JSON or similar.  I can get the updates in by tomorrow morning.

File: rsv-metrics/libexec/probes/gratia-config-probe [hunk 2]
```diff
@@ -268,77 +269,107 @@ EOF
...
-  }
-
-  # This means either the collector didn't respond or we've checked in at least once.
-  $probe_contact_check = ParseDate($probe_contact_check);
-  my $last_contact_string = UnixDate($probe_contact_check, "%Y-%m-%d %T %Z");
-  my $cmpDateHour = DateCalc("today", "- 1 hour");
-  my $cmpDateDay = DateCalc("today", "- 1 day");
-  my $probe_seen_by_soaphost;
-  if (!$probe_contact_check) {
+  if ($probe_contact_check =~ /"hits.*?total"\s*:\s*(\d+)/) {
```
Comment (edquist) [inline rsv-metrics/libexec/probes/gratia-config-probe]: Thanks @sthapa, the `JSON` addition is great.  Would it be possible to parse `$probe_contact_check` using `decode_json` also?

**Comment by sthapa** [comment]
Sure, I'll submit the changes later today.

**Comment by sthapa** [comment]
Just committed changes.  Incidentally, perl is awful.  If I wanted to use the dereference operator ( -> ) so many times in a row, I would have been writing c++ or c.

**Comment by edquist** [comment]
>Incidentally, perl is awful

Haha... I'm sorry.  But I think you can do
```$$response{hits}{hits}[0]{_source}{received}```
instead of
```$response->{hits}->{hits}[0]->{_source}->{received}```

**Comment by sthapa** [comment]
Unfortunately, I don't think I can.  The decode_json function returns references to hashes or arrays instead of a hash or array.  Since I think it does some sort of recursive parsing, it's references if the item isn't an atomic data type.

**Comment by edquist** [comment]
>  The decode_json function returns references to hashes or arrays instead of a hash or array

I don't think this is a problem.  In perl all the nested data structures are like that: nested hash and array references - but once you have a scalar which holds a reference to a hash or array, you can just tack on `{key}` or `[idx]` things which will automatically dereference.

For example:

```
>> $x = { hits => { hits => [ { _source => { received => "2018-01-31T23:08:13.389Z" } } ] } }
=> HASH(0x11d43c8)
>> $$x{hits}{hits}[0]{_source}{received}
=> 2018-01-31T23:08:13.389Z
```
I think the key thing to notice is that your initial `$response` variable is a scalar (thus the `$`) that holds a `HASH` ref.  So to do the first dereference you either need `$response->` or `$$response`.  After that you can put all the `{key}`, `[idx]` things you want in a row.

In particular, the thing you _can't_ do is `$response{hits}`, because that would be referring to the hash `%response`, which is different from the scalar `$response` where you stored your `HASH` ref returned from `decode_json`.

**Comment by edquist** [comment]
As for things relevant to the review, I noticed that the hash under `$$response{hits}{hits}[0]{_source}` has a `@received` key, but no `received` key, so I think that needs to be changed.

Eg, here:

```
>> $response->{hits}->{hits}[0]->{_source}->{received}
=> 
>> $response->{hits}->{hits}[0]->{_source}->{'@received'}
=> 2018-01-31T23:08:13.389Z
>> $$response{hits}{hits}[0]{_source}{'@received'}
=> 2018-01-31T23:08:13.389Z
```

... The only other thing is I'd like to do the same json encode/decode for the second query (the one with probename + sitename)

**Comment by edquist** [comment]
@sthapa, thanks for figuring out how to do these queries against gracc and how to use the perl json library!  Let me know if you are ok with the additional updates i've pushed to address my above comments.

**Comment by sthapa** [comment]
Hi,

I actually did a replace on line 303 to change the fieldname from @received to received.  It made accessing the field once the json is decoded a bit easier.  I can remove the string replacement if you'd prefer.  

Can you give me the line for the second query?  I can't find it.

**Comment by edquist** [comment]
> I actually did a replace on line 303 to change the fieldname from `@received` to `received`. It made accessing the field once the json is decoded a bit easier

Ah!  Now i see it!  (Sorry, i was testing fragments of the code and didn't include that part.)  It's fine the way you had it, I've removed my commit that that changed the check to `@received`.

> Can you give me the line for the second query? I can't find it.

The second query was on [line 358 of your last commit](https://github.com/opensciencegrid/rsv/pull/19/commits/467044f910d7fbb7837f786558d8179099455220#diff-08b6175f83e41b1c4327276095c05ecbR358), and I use the json encoded/decode functions for it in [my last commit](https://github.com/opensciencegrid/rsv/pull/19/commits/2ea1c7b0d0dad61e58093a226fd41c0b0522a6c9#diff-08b6175f83e41b1c4327276095c05ecbL358).

Let me know if it looks ok or if you see any issues...

**Comment by sthapa** [comment]
Ah, I'll check in updates for this later today.

**Comment by sthapa** [comment]
Ah, everything looks good.  I tried using the code to on a working install with some hardcoded probename and site values and everything seems to be working.

**Comment by edquist** [comment]
Great!  Thanks @sthapa!

**Pull request merged**